packages/compiler-cli/linker/src/file_linker/partial_linkers/partial_directive_linker_1.ts

```diff
@@ -154,6 +157,72 @@ function toQueryMetadata<TExpression>(obj: AstObject<R3DeclareQueryMetadata, TEx
   };
 }
 
+/**
+ * Derives the host directives structure from the AST object.
+ */
+function toHostDirectivesMetadata<TExpression>(
+    hostDirectives: AstValue<
+        (o.Expression | R3DeclareHostDirectiveMetadata)[]|o.FunctionExpr|undefined, TExpression>):
```

Comment:
Partially compiled host directive declarations should _always_ be an array of `R3DeclareHostDirectiveMetadata`, so we should be able to drop the `o.Expression` and `o.FunctionExpr` form:
```suggestion
        R3DeclareHostDirectiveMetadata[]|undefined, TExpression>):
```

The forward reference should be tracked individually (the `directive` field of each one might be a function).